agents/agents-ext/src/commonMain/kotlin/ai/koog/agents/ext/tool/search/RegexSearchTool.kt

```diff
@@ -0,0 +1,154 @@
+        pattern: String,
+        limit: Int,
+        skip: Int,
+        caseSensitive: Boolean,
+        linesAroundSnippet: Int = 2,
+    ): Flow<FileSystemEntry.File> {
+        val options = mutableSetOf<RegexOption>()
+        if (!caseSensitive) options.add(RegexOption.IGNORE_CASE)
+
+        return searchByRegex(
+            fs = fs,
+            start = path,
+            regex = Regex(pattern, options)
+        )
+            .drop(skip)
+            .take(limit)
+            .mapNotNull { match ->
+                val content = fs.readText(match.file)
+                val snippets = match.ranges.map { range ->
+                    val extended = extendRangeByLines(content, range, linesAroundSnippet, linesAroundSnippet)
+                    FileSystemEntry.File.Content.Excerpt.Snippet(
+                        text = extended.substring(content),
+                        range = extended
+                    )
+                }
+                if (snippets.isEmpty()) return@mapNotNull null
+                val metadata = fs.metadata(match.file) ?: return@mapNotNull null
+                val contentType = fs.getFileContentType(match.file)
+                FileSystemEntry.File(
+                    name = fs.name(match.file),
+                    extension = fs.extension(match.file),
+                    path = fs.toAbsolutePathString(match.file),
+                    hidden = metadata.hidden,
+                    size = buildFileSize(fs, match.file, contentType),
+                    contentType = contentType,
+                    content = FileSystemEntry.File.Content.Excerpt(snippets)
+                )
+            }
+    }
+
+    /**
+     * A match of one file and the ranges within it that matched a regex.
+     */
+    private data class ContentMatch<Path>(val file: Path, val ranges: List<DocumentProvider.DocumentRange>)
+
+    /**
+     * Recursively searches starting at [start] for text files whose contents match [regex].
+     * Returns a flow of [ContentMatch] where each item corresponds to a file and its matched ranges.
+     */
+    private fun <Path> searchByRegex(
+        fs: FileSystemProvider.ReadOnly<Path>,
+        start: Path,
+        regex: Regex
+    ): Flow<ContentMatch<Path>> = flow {
+        when (fs.metadata(start)?.type) {
+            FileMetadata.FileType.File -> {
+                try {
+                    if (fs.getFileContentType(start) != FileMetadata.FileContentType.Text) return@flow
+                    val content = fs.readText(start)
+                    val ranges = regex.findAll(content).map { mr ->
+                        val s = mr.range.first
+                        val e = mr.range.last + 1 // exclusive
+                        DocumentProvider.DocumentRange(s.toPosition(content), e.toPosition(content))
+                    }.toList()
+                    if (ranges.isNotEmpty()) emit(ContentMatch(start, ranges))
+                } catch (e: kotlinx.coroutines.CancellationException) {
```

Comment:
Please make it a top-level import. Also, AFAIK it's better to use `kotlin.coroutines.cancellation.CancellationException`